Remove empty fields from generated configs

mayankshah1607 opened the pull request:
Fixes
- https://github.com/linkerd/linkerd2/issues/2962
- https://github.com/linkerd/linkerd2/issues/2545

### Problem
Field omissions for workload objects are not respected while marshaling to JSON.

### Solution
After digging a bit into the code, I came to realize that while marshaling, workload objects have empty structs as values for various fields which would rather be omitted. As of now, the standard library`encoding/json` does not support zero values of structs with the `omitemty` tag. The relevant issue can be found [here](https://github.com/golang/go/issues/11939). To tackle this problem, the object declaration should have _pointer-to-struct_ as a field type instead of _struct_ itself. However, this approach would be out of scope as the workload object declaration is handled by the k8s library.

I was able to find a drop-in replacement for the `encoding/json` library which supports zero value of structs with the `omitempty` tag. It can be found [here](https://github.com/clarketm/json). I have made use of this library to implement a simple filter like functionality to remove empty tags once a YAML with empty tags is generated, hence leaving the previously existing methods unaffected

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

Pothulapati commented on pkg/inject/inject.go:
```diff
@@ -283,7 +288,9 @@ func (conf *ResourceConfig) JSONToYAML(bytes []byte) ([]byte, error) {
 	}
-	return yaml.Marshal(obj)
+
+	filteredJSON, _ := jsonfilter.Marshal(&obj)
```
Thanks for the contribution :) 
Add a check for err.
Also, Can you move this Conversion functionality into a common function and re-use that?

mayankshah1607 replied:
@Pothulapati Sure, will do! If this approach seems fine, I could go ahead and clean up the code a bit! Just needed a thumbs-up on this kind of implementation

adleong commented:
It looks like https://github.com/clarketm/json pulls in commits from https://github.com/golang/go master.  Currently it's based off of https://github.com/golang/go/commit/54697702e435bddb69c0b76b25b3209c78d2120a which is the second latest commit.  So it's actually using a newer version of `encoding/json` than we are (1.13.4).  

The relevant diff is effectively this commit: https://github.com/clarketm/json/commit/6acb9d8d1d575a7961e9414c3eb3b3237bfb2bb4 which looks pretty sane to me.

So I'm pretty comfortable with adding this dependency.

adleong commented on pkg/inject/inject.go:
```diff
@@ -177,7 +178,11 @@ func (conf *ResourceConfig) AppendPodAnnotation(k, v string) {
 // YamlMarshalObj returns the yaml for the workload in conf
 func (conf *ResourceConfig) YamlMarshalObj() ([]byte, error) {
-	return yaml.Marshal(conf.workload.obj)
+	j, err := jsonfilter.Marshal(conf.workload.obj)
```
This should definitely have a comment explaining why we're not using `encoding/json` here.

mayankshah1607 commented:
@adleong @Pothulapati done! Mind taking a look now?
I've added a reusable method `getFilteredJSON()` and made it work with the existing code for `inject` / `uninject`. I've also added a comment explaining why we're not making use of `encoding/json` in the above mentioned method, as @adleong mentioned :smile:

adleong commented on pkg/inject/inject.go:
```diff
@@ -939,5 +949,22 @@ func (conf *ResourceConfig) InjectNamespace(annotations map[string]string) ([]by
+//empty structs are respected by `omitempty` tags. We make use of a drop-in
+//replacement of the standard json/encoding library, without which empty struct values
+//present in workload objects would make it into the marshaled JSON.
+func getFilteredJSON(conf runtime.Object) ([]byte, error) {
```
I think the `getFilteredJSON` method you have have here is identical to `jsonfilter.Marshal`.  It's great that we have this comment, but I don't think the function gives us any benefit.

mayankshah1607 replied:
@adleong We'd have to be making use of the `jsonfilter.Marshal` method at multiple places. I wasn't really sure if it would be a nice idea to add such a comment at multiple places. I thought it would be a great idea to just have a reusable function that makes use of `jsonfilter.Marshal` that contributors could use in the future, instead of having to directly deal with a different version of `jso/encoding` by themselves, preserving the modularity of the codebase. 

WDYT?

adleong replied:
Gotcha.  In that case, I think the `getFilteredJSON` function body can simply be a one-liner: `return jsonfilter.Marshal(&conf)`

mayankshah1607 commented:
@adleong could you take a look now? I hope it looks fine :smile:

mayankshah1607 commented:
Thank you @adleong !
@cpretzer mind taking a look do that we could get this merged?
